server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java

```diff
@@ -0,0 +1,184 @@
+
+    /**
+     * A histogram tracking the sizes of batches that we have processed.
+     */
+    private final Histogram batchSizeHist = KafkaYammerMetrics.defaultRegistry()
+            .newHistogram(batchSizeHistName, true);
+
+    private final AtomicReference<MetadataProvenance> lastAppliedImageProvenance = new AtomicReference<>(MetadataProvenance.EMPTY);
+    private final AtomicLong metadataLoadErrorCount = new AtomicLong(0);
+    private final AtomicLong metadataApplyErrorCount = new AtomicLong(0);
+
+    private final Metrics metrics;
+    private final MetricName lastAppliedRecordOffsetName;
+    private final MetricName lastAppliedRecordTimestampName;
+    private final MetricName lastAppliedRecordLagMsName;
+    private final MetricName metadataLoadErrorCountName;
+    private final MetricName metadataApplyErrorCountName;
+
+    public BrokerServerMetrics(Metrics metrics) {
+        this.metrics = metrics;
+        lastAppliedRecordOffsetName = metrics.metricName(
+                "last-applied-record-offset",
+                METRIC_GROUP_NAME,
+                "The offset of the last record from the cluster metadata partition that was applied by the broker"
+        );
+        lastAppliedRecordTimestampName = metrics.metricName(
+                "last-applied-record-timestamp",
+                METRIC_GROUP_NAME,
+                "The timestamp of the last record from the cluster metadata partition that was applied by the broker"
+        );
+        lastAppliedRecordLagMsName = metrics.metricName(
+                "last-applied-record-lag-ms",
+                METRIC_GROUP_NAME,
+                "The difference between now and the timestamp of the last record from the cluster metadata partition that was applied by the broker"
+        );
+        metadataLoadErrorCountName = metrics.metricName(
+                "metadata-load-error-count",
+                METRIC_GROUP_NAME,
+                "The number of errors encountered by the BrokerMetadataListener while loading the metadata log and generating a new MetadataDelta based on it."
+        );
+        metadataApplyErrorCountName = metrics.metricName(
+                "metadata-apply-error-count",
+                METRIC_GROUP_NAME,
+                "The number of errors encountered by the BrokerMetadataPublisher while applying a new MetadataImage based on the latest MetadataDelta."
+        );
+
+        addMetric(metrics, lastAppliedRecordOffsetName, ignored -> lastAppliedImageProvenance.get().lastContainedOffset());
+        addMetric(metrics, lastAppliedRecordTimestampName, ignored -> lastAppliedImageProvenance.get().lastContainedLogTimeMs());
+        addMetric(metrics, lastAppliedRecordLagMsName, now -> now - lastAppliedImageProvenance.get().lastContainedLogTimeMs());
+        addMetric(metrics, metadataLoadErrorCountName, ignored -> metadataLoadErrorCount.get());
+        addMetric(metrics, metadataApplyErrorCountName, ignored -> metadataApplyErrorCount.get());
```

Comment:
For these 5 lines, could we do:
```
metrics.addMetric(lastAppliedRecordOffsetName, (config, now) -> lastAppliedImageProvenance.get().lastContainedOffset());
metrics.addMetric(lastAppliedRecordTimestampName, (config, now) -> lastAppliedImageProvenance.get().lastContainedLogTimeMs());
metrics.addMetric(lastAppliedRecordLagMsName, (config, now) -> now - lastAppliedImageProvenance.get().lastContainedLogTimeMs());
metrics.addMetric(metadataLoadErrorCountName, (config, now) -> metadataLoadErrorCount.get());
metrics.addMetric(metadataApplyErrorCountName, (config, now) -> metadataApplyErrorCount.get());
```
and not need the `FuncGauge` class and the `addMetrics()` method?

Reply:
that's a good point, done.